File: app/src/main/java/com/dialogapp/dialog/ui/common/PostViewHolder.java
package com.dialogapp.dialog.ui.common;

import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.text.Html;
import android.text.Spannable;
import android.text.method.LinkMovementMethod;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.RequestManager;
import com.bumptech.glide.request.RequestOptions;
import com.dialogapp.dialog.R;
import com.dialogapp.dialog.model.Item;
import com.dialogapp.dialog.ui.profilescreen.ProfileActivity;
import com.dialogapp.dialog.util.GlideImageGetter;

import java.util.regex.Pattern;

import butterknife.BindView;
import butterknife.ButterKnife;

import static android.text.Html.FROM_HTML_MODE_LEGACY;
import static android.util.TypedValue.COMPLEX_UNIT_DIP;
import static android.util.TypedValue.applyDimension;

public class PostViewHolder extends BaseViewHolder<Item> {
    private Context context;
    private Pattern pattern;

    @BindView(R.id.image_thumbnail)
    ImageView thumbnail;

    @BindView(R.id.text_username)
    TextView username;

    @BindView(R.id.text_content)
    TextView content;

    @BindView(R.id.text_time)
    TextView time;

    public PostViewHolder(View itemView, RequestManager glide) {
        super(itemView, glide);
        ButterKnife.bind(this, itemView);

        this.context = this.itemView.getContext();
        this.pattern = Pattern.compile("<img");
    }

    @Override
    public void onBind(Item item, int position) {
        thumbnail.setOnClickListener(view -> {
            Intent intent = new Intent(context, ProfileActivity.class);
            intent.putExtra(ProfileActivity.EXTRA_USERNAME, item.author.microblog.username);
            context.startActivity(intent);
        });

        glide.load(item.author.avatar)
                .apply(RequestOptions.circleCropTransform())
                .into(thumbnail);

        username.setText(item.author.microblog.username);
        time.setText(item.microblog.dateRelative);

        bindHtmlContent(item.contentHtml);
    }

    private void bindHtmlContent(String contentHtml) {
        CharSequence spannedText;
        if (pattern.matcher(contentHtml).find()) {
            float size = applyDimension(COMPLEX_UNIT_DIP, 64, context.getResources().getDisplayMetrics());
            GlideImageGetter imageGetter = new GlideImageGetter(glide, content, (int) size);
            spannedText = getSpanned(contentHtml, imageGetter);
        } else {
            spannedText = getSpanned(contentHtml, null);
        }

        Spannable htmlString = (Spannable) spannedText;
        LinkClickHandler.makeLinksClickable(context, htmlString);
        content.setText(trimTrailingWhitespace(spannedText));
        content.setMovementMethod(LinkMovementMethod.getInstance());
    }

    private CharSequence getSpanned(String contentHtml, GlideImageGetter imageGetter) {
        if (Build.VERSION.SDK_INT >= 24) {
            return Html.fromHtml(contentHtml, FROM_HTML_MODE_LEGACY, imageGetter, null);
        } else {
            return Html.fromHtml(contentHtml, imageGetter, null);
        }
    }

    public void clearView() {
        glide.clear(thumbnail);
        username.setText(null);
        time.setText(null);
        GlideImageGetter.clear(content);
    }

    /**
     * Remove unwanted newlines added by Html.fromHtml
     *
     * @param source
     * @return
     */
    private CharSequence trimTrailingWhitespace(CharSequence source) {
        if (source == null)
            return "";

        int i = source.length();

        // loop back to the first non-whitespace character
        while (--i >= 0 && Character.isWhitespace(source.charAt(i))) {
        }

        return source.subSequence(0, i + 1);
    }
}

File: app/src/main/java/com/dialogapp/dialog/ui/common/PostViewHolder.kt
package com.dialogapp.dialog.ui.common

import android.os.Bundle
import android.view.View
import androidx.recyclerview.widget.RecyclerView
import com.dialogapp.dialog.GlideRequests
import com.dialogapp.dialog.R
import com.dialogapp.dialog.databinding.PostItemBinding
import com.dialogapp.dialog.model.Post
import com.dialogapp.dialog.ui.util.GlideImageGetter
import com.dialogapp.dialog.ui.util.HtmlTextHelper
import com.dialogapp.dialog.ui.util.ImageGetterOptions

class PostViewHolder(private val view: View, val binding: PostItemBinding, private val glide: GlideRequests,
                     private val postClickedListener: PostClickedListener, private val imageGetterOptions: ImageGetterOptions)
    : RecyclerView.ViewHolder(view) {

    fun bind(post: Post?) {
        binding.textFullname.text = post?.author?.name
        binding.textUsername.text = String.format(view.resources.getString(R.string.post_username),
                post?.author?.microblog?.username)
        binding.textTime.text = String.format(view.resources.getString(R.string.post_time),
                post?.microblog?.dateRelative)
        HtmlTextHelper(glide, postClickedListener, imageGetterOptions, post?.contentHtml)
                .setHtmlContent(binding.textContent)
        glide.load(post?.author?.avatar).into(binding.imageThumbnail)

        binding.buttonReply.setOnClickListener {
            postClickedListener.onReplyClicked(post?.id, post?.author?.microblog?.username, post?.contentHtml)
        }

        binding.buttonConv.visibility = when (post?.microblog?.isConversation) {
            true -> {
                binding.buttonConv.setOnClickListener {
                    postClickedListener.onConversationButtonClicked(post.id)
                }
                View.VISIBLE
            }
            else -> {
                View.INVISIBLE
            }
        }

        when (post?.microblog?.isFavorite) {
            true -> binding.buttonFav.setImageResource(R.drawable.ic_round_favorite_24px)
            else -> binding.buttonFav.setImageResource(R.drawable.ic_round_favorite_border_24px)
        }
        binding.buttonFav.setOnClickListener {
            postClickedListener.onFavoriteButtonClicked(post?.id, post?.belongsToEndpoint)
        }

        binding.imageThumbnail.setOnClickListener {
            postClickedListener.onProfileClicked(post?.author?.microblog?.username!!)
        }

        binding.buttonOverflow.setOnClickListener {
            if (post != null) {
                postClickedListener.onOverflowMenuClicked(post.id, post.url,
                        post.author.microblog?.username!!, post.microblog.isDeletable)
            }
        }
    }

    fun updatePost(payloads: MutableList<Any>, post: Post?) {
        val diffBundle = payloads[0] as Bundle

        for (key in diffBundle.keySet()) {
            when (key) {
                PK_CONVERSATION -> {
                    when (diffBundle.getBoolean(PK_CONVERSATION)) {
                        true -> {
                            binding.buttonConv.visibility = View.VISIBLE
                            binding.buttonConv.setOnClickListener {
                                if (post?.id != null)
                                    postClickedListener.onConversationButtonClicked(postId = post.id)
                            }
                        }
                        false -> {
                            binding.buttonConv.visibility = View.INVISIBLE
                        }
                    }
                }
                PK_FAVORITE -> {
                    when (diffBundle.getBoolean(PK_FAVORITE)) {
                        true -> binding.buttonFav.setImageResource(R.drawable.ic_round_favorite_24px)
                        else -> binding.buttonFav.setImageResource(R.drawable.ic_round_favorite_border_24px)
                    }
                }
                PK_DATE -> {
                    binding.textTime.text = diffBundle.getString(PK_DATE)
                }
                PK_AVATAR -> {
                    glide.load(diffBundle.getString(PK_AVATAR))
                            .into(binding.imageThumbnail)
                }
                PK_USERNAME -> {
                    binding.textUsername.text = diffBundle.getString(PK_USERNAME)
                }
                PK_CONTENT -> {
                    HtmlTextHelper(glide, postClickedListener, imageGetterOptions, diffBundle.getString(PK_CONTENT))
                            .setHtmlContent(binding.textContent)
                }
            }
        }
    }

    fun recycle() {
        binding.textFullname.text = null
        binding.textUsername.text = null
        binding.textTime.text = null
        binding.textContent.text = null

        glide.clear(binding.imageThumbnail)
        GlideImageGetter.clear(binding.textContent)
    }
}